### DISCUSSION TITLE
How do I choose sane variable defaults for ContextDatabase?

### DISCUSSION THREAD

**Discussion by liona24** [NationalSecurityAgency/ghidra#6454] [Q&A]
Hey. I am currently trying to use the decompiler as a library and I have a simple binary to pcode decompiler working ( following the [sleigh.hh docs](https://github.com/NationalSecurityAgency/ghidra/blob/master/Ghidra/Features/Decompiler/src/decompile/cpp/sleigh.hh) ).

I am currently using a simple `x86:LE:64:default` target with the shipped `ghidra_11.0.3_PUBLIC/Ghidra/Processors/x86/data/languages/x86-64.sla` SLA file (uncompressed).

I noticed that the generated pcode is complete garbage for some sample x64 code unless I specify the following defaults for my `Sleigh` class:
```cpp
      sleigh.setContextDefault("addrsize", 2);
      sleigh.setContextDefault("opsize", 2);
      sleigh.setContextDefault("longMode", 1);
```
All of them are `0` by default. I am not exactly sure what those values mean and how I am supposed to set them. The header docs mention that `"addrsize" == 1` means 32 bit mode, so I just set it to `2` for 64 bit.

There are quite a few more of those variables (?) present in the .SLA file, and since the defaults were not working properly I am uncertain whether the default is fine for the rest of them (`bit64`, `protectedMode`, `vexMode`, .. to name a few).

This seems very specific to x86-64 as well, how do I manage other targets?

Is there a good way to initialize them? Grepping through the codebase, they do not ever seem to be set by the java code. So how are they supposed to be managed?

Thanks!

**Comment by ghidra1** [comment]
Each unique language ID (e.g., `x86:LE:64:default`) is defined by an `*.ldefs` file which also matches-up to a `*.pspec` file.  With this `*.pspec` the default context state may be adjusted for a specific language (e.g., `x86-64.pspec`)
